### PR TITLE
feat: add `scroll_getTxBlockTraceOnTopOfBlock ` API

### PR DESCRIPTION
**Pull request by 0xmountaintop**
## 1. Purpose or design rationale of this PR

Currently, if a tx encounters ccc overflow and is skipped, we can only use `debug_traceCall` to replay and get the `structLogs`. However, a tx's `structLogs` is insufficient, or at least inconvenient, to analyze its ccc row consumption. This API add support for getting a wrapped blocktrace for a call/tx, at a given block's state.

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [X] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [X] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [ ] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: internal/web3ext/web3ext.go [hunk 1]
```diff
@@ -865,6 +865,12 @@ web3._extend({
 			call: 'scroll_getBlockTraceByNumberOrHash',
 			params: 1
 		}),
+		new web3._extend.Method({
+			name: 'getBlockTraceForCallAt',
+			call: 'scroll_getBlockTraceForCallAt',
+			params: 3,
+			inputFormatter: [null, null, null]
```
Comment (colinlyguo) [inline internal/web3ext/web3ext.go]:
```suggestion
			params: 3
```

File: eth/tracers/api_blocktrace.go [hunk 2]
```diff
@@ -30,20 +32,51 @@ func (api *API) GetBlockTraceByNumberOrHash(ctx context.Context, blockNrOrHash r
...
+	}
+
+	block = types.NewBlockWithHeader(block.Header()).WithBody([]*types.Transaction{args.ToTransaction()}, nil)
+
+	var traceConfig *TraceConfig
+	var stateOverrides *ethapi.StateOverride
+	if config != nil {
+		traceConfig = &TraceConfig{
+			LogConfig: config.LogConfig,
+			Tracer:    config.Tracer,
```
Comment (colinlyguo) [inline eth/tracers/api_blocktrace.go]: Would this trigger `log.Warn("Tracer params is unsupported")`?

Comment (colinlyguo) [reply eth/tracers/api_blocktrace.go]: btw, why `Tracer` params is unsupported?

File: eth/tracers/api_blocktrace.go [hunk 3]
```diff
@@ -30,20 +32,51 @@ func (api *API) GetBlockTraceByNumberOrHash(ctx context.Context, blockNrOrHash r
...
+		return nil, errors.New("invalid arguments; neither block nor hash specified")
+	}
+	if err != nil {
+		return nil, err
+	}
+	if block.NumberU64() == 0 {
+		return nil, errors.New("genesis is not traceable")
+	}
+
+	block = types.NewBlockWithHeader(block.Header()).WithBody([]*types.Transaction{args.ToTransaction()}, nil)
```
Comment (Thegaram) [inline eth/tracers/api_blocktrace.go]: So we actually replay and trace the transaction on top of `block.parent`'s state, right?

Comment (0xmountaintop) [reply eth/tracers/api_blocktrace.go]: yes, see `(api *API) createTraceEnv`.

we apply a mock block, containing the given tx/call, on top of the parent block

Comment (0xmountaintop) [reply eth/tracers/api_blocktrace.go]: so your point is, the API name is confusing? I agree. Let me think of a better name and add more doc

Comment (0xmountaintop) [reply eth/tracers/api_blocktrace.go]: what about `scroll_getBlockTraceForCallAtParentBlock`?

File: internal/web3ext/web3ext.go (hunk 1 above)
Comment (0xmountaintop) [reply internal/web3ext/web3ext.go]: Here I just follow
```
		new web3._extend.Method({
			name: 'traceCall',
			call: 'debug_traceCall',
			params: 3,
			inputFormatter: [null, null, null]
		}),
```

So I think it's unnecessary to delete this line.

File: eth/tracers/api_blocktrace.go (hunk 2 above)
Comment (0xmountaintop) [reply eth/tracers/api_blocktrace.go]: we don't want user to specify tracer here, we only want them to use `vm.NewStructLogger(env.logConfig)`

see https://github.com/scroll-tech/go-ethereum/blob/65a3a7fc238954d48cb47d83e4b676e98dc52ac8/core/trace.go#L272-L274

File: eth/tracers/api_blocktrace.go (hunk 3 above)
Comment (Thegaram) [reply eth/tracers/api_blocktrace.go]: I think as long as the behaviour is consistent with `debug_traceCall` then it's okay. Btw you can also consider naming it `scroll_traceCall`.

File: internal/web3ext/web3ext.go (hunk 1 above)
Comment (colinlyguo) [reply internal/web3ext/web3ext.go]: Oh I see.

File: eth/tracers/api_blocktrace.go (hunk 2 above)
Comment (colinlyguo) [reply eth/tracers/api_blocktrace.go]: Oh, that's reasonable.

File: eth/tracers/api_blocktrace.go (hunk 3 above)
Comment (0xmountaintop) [reply eth/tracers/api_blocktrace.go]: then it's different.


`debug_traceCall` traces at the given block
but this new API traces at the parent block

**Review by lispc — APPROVED** [review]

**Pull request merged**